### PR TITLE
Added a new EnforceRule for checking the licensing definition

### PR DESCRIPTION
**Pull request by recena**
This PR is an alternative to https://github.com/jenkinsci/plugin-pom/pull/85 and it is downstream of https://github.com/mojohaus/extra-enforcer-rules/pull/34

@jenkinsci/code-reviewers and previous reviewers in #85 @stephenc @abayer @oleg-nenashev @daniel-beck ....

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -544,7 +545,7 @@
           <dependency>
             <groupId>org.codehaus.mojo</groupId>
             <artifactId>extra-enforcer-rules</artifactId>
-            <version>1.0-beta-4</version>
+            <version>1.0-beta-7-SNAPSHOT</version>
```
Comment (recena) [inline pom.xml]: I need to know how to deploy this snapshot /cc @stephenc

Comment (batmat) [reply pom.xml]: @recena anyone who is a maintainer of any plugin can release a SNAPSHOT of any other. Just do `mvn clean deploy`.

(IIUC/IIRC, it's just that the handle must appear somewhere in https://github.com/jenkins-infra/repository-permissions-updater/ permissions, amirite @daniel-beck?)

Comment (batmat) [reply pom.xml]: Oh wait, this is extra-enforcer-rule, hence MojoHaus, not Jenkins org. So I can do it too if really needed. We don't have the same policy there, though.

Comment (recena) [reply pom.xml]: @batmat I would appreciate it.

**Comment by jglick** [comment]
Anyone in the Jenkins org can deploy a _snapshot_ to the _Jenkins_ repo:

```bash
mvn clean install source:jar deploy:deploy -DaltDeploymentRepository=maven.jenkins-ci.org::default::https://repo.jenkins-ci.org/snapshots/
```

Then specify the _timestamped version_ in this PR for testing.

If you want to deploy a _beta release_ to the Jenkins repo (which would be a prerequisite for this PR being merged, but assumes that the upstream PR has already been merged), you would need special permissions. But since this change is not at all critical, I would probably just want for the upstream change to be officially released.

**Comment by jglick** [comment]
> We should be requiring `licenses` for `HOSTING` requests and for release to `repo.jenkins-ci.org`

Even deployment to `repo.jenkins-ci.org` per se does not matter. IMO it is the job of `backend-update-center2` to enforce this kind of metadata. If your plugin has bad metadata, it just gets yanked from the UC until you fix it.

**Comment by daniel-beck** [comment]
> If your plugin has bad metadata, it just gets yanked from the UC until you fix it.

Would be possible. At least it's not an external dependency, like the wiki was, so the only thing that could break this would be bad plugin updates. Would need to be implemented on a per-version level, so that a plugin release with a bad license would be ignores in terms of eligibility for distribution, and make it fall back to a previous one, if any.

> they would not build without a license, yet we do not wish to push a license onto sample code which people should be free to copy and paste into whatever they like.

Treat like the commented-out `distributionManagement`? Perhaps even the same comment?

**Comment by jglick** [comment]
> Would need to be implemented on a per-version level, so that a plugin release with a bad license would be ignores in terms of eligibility for distribution, and make it fall back to a previous one, if any.

Exactly.

> Treat like the commented-out `distributionManagement`?

`scm` I guess you mean. Yes.

**Comment by recena** [comment]
This is so bored.

**Comment by amuniz** [comment]
Having plugins without any kind of license is a problem. From https://choosealicense.com/no-license :

```
If you find software that doesn’t have a license, that generally 
means you have no permission from the creators of the software to 
use, modify, or share the software. Although a code host such as 
GitHub may allow you to view and fork the code, this does not imply 
that you are permitted to use, modify, or share the software for any purpose.
```

IMO the enforcer to require a license is ok. For `jenkinsci/archetypes` I think it is ok to use MIT (the recommended one for Jenkins plugins in wiki IIRC).

The parent pom is not written in stone, it can be easily overridden on children projects which can opt-out of the default licensing.

**Comment by stephenc** [comment]
The plugin parent pom is not just for plugins being hosted on Jenkins infra.

As such, it is entirely incorrect to enforce this here in this way as it breaks existing use cases.

Just because the in-house closed source developer of plugins is never visible to you does not mean you get to break their build.

In some cases the person responsible for maintenance of the plugin can be a junior dev who was landed with it once the original developer left. They are likely not in a position to determine or negotiate the licensing, so forcing them to specify is a bad thing.

I have expressed my opinion, namely enforce at UC, use hpi plugin to validate pre-publishing. The enforcer does not have the smarts to turn off checks where `distributionManagement` is not on Jenkins.io servers, which is what would be required for checks to be part of this pom.

This pom has an implicit API contract, namely that you can keep upgrading to get better functionality. Mandatory changes to the pom that are forced by upgrading this pom, especially on users for whom those mandatory changes are irrelevant is a very bad break if that API

**Comment by recena** [comment]
Once again, I have a proposal with three blockers.

Does it make sense to continue working here?

**Comment by stephenc** [comment]
The idea is ok, the approach is, IMHO, incorrect.

Start with adding a goal to the hpi plugin to validate the pom for OSS hosting, the goal should auto-skip if distributionManagement is to a hostname that is neither Jenkins-CI.org nor Jenkins.io

It can fail the build if hosting on Jenkins infra and there is no license or no wiki url.

**Comment by abayer** [comment]
I agree with @stephenc. I’m not sure exactly where we want to hook that goal into the lifecycle, but I’d assume we only want to run it as part of `release:prepare` - I know from experience that not every internal plugin ever bothers updating `distributionManagement` and may only be built as snapshots, so we don’t want to be doing this for a simple `mvn clean install`.

**Comment by stephenc** [comment]
We want to bind it to the `verify` phase (and OT but some of the current enforcer rules need to be moved to the verify phase too... i’m looking at you require-upper-bounds) but this should be bound only in the release activated profile so that `release:prepare` picks it up

**Comment by stephenc** [comment]
Note: If only bound to the release profile, we could alternatively bind it early, eg to `validate`

Alternatively, it could be fatal only for release version numbers and live bound to `verify` in the main lifecycle

**Comment by jglick** [comment]
I still prefer doing this check in `backend-update-center2` or whatever it is, since that is the point at which we actually care to enforce site policies like this. Deployment to Artifactory _per se_ is irrelevant, and may in fact have come from a Gradle build—or from a plugin which simply declined to update its parent, which is going to be a lot.

That said, an Enforcer rule activated only when `distributionManagement` points to `https://repo.jenkins-ci.org/releases/` _and_ the `jenkins-release` profile is active would be better than nothing. That would avoid breaking builds for people who are doing non-release builds without bothering to modify `distributionManagement`, or for people deploying releases to other servers. It would at least enforce our policy for the case of people doing ongoing Maven releases with a new parent POM, covering some slowly growing fraction of the plugins in the update center.

**Comment by abayer** [comment]
I think we've established a consensus here - an enforcer rule that only applies for releases to the Jenkins repo seems to be the best option.

**Comment by jglick** [comment]
Again, I do not consider that the “best” option, just better than nothing at all. If I correctly understand @stephenc’s argument, he agrees that actual enforcement belongs in the update center, but was suggesting _also_ having prevalidation for users of the newest Maven parent as a convenience.

**Comment by abayer** [comment]
Yes, sorry, I forgot to mention the UC generation aspect of it as well. =)

**Comment by recena** [comment]
If my opinion wants to be considered, the idea about using the UCC as gatekeeper is not appropriate. Taking into account the current version of that service. My main reason were shared here https://github.com/jenkinsci/plugin-pom/pull/85#issuecomment-340621551.

**Comment by stephenc** [comment]
@jglick it *cannot* be an enforcer rule as enforcer does not let us toggle rules based on distributionManagement. It *should* be a validation goal on the hpi plugin.

Correct that my view is:

1. Enforce on UC
2. Assist user comprehension in pom.

As #2 is easier to start from, I recommend starting there

**Comment by stephenc** [comment]
@recena, the UC *is* the ideal place to enforce. You are arguing a different thing, namely getting to the point where the UC enforces this and plugin devs understanding as to why publishing blocked, is a big bit of work.... IOW “today the UC is not ideal” != “in a perfect world we would do this in the UC”

I am saying today, start with `hpi:validate-jenkins-hosting` and then we see about the UC changes after that

**Comment by recena** [comment]
@stephenc as you said it, _in a perfect world..._ I try to provide solutions taking into account the real circumstances. The option that you are considering was not feasible for me at this moment.

**Comment by recena** [comment]
In meantime, and until the right solution arrives, we have an issue. IMHO a critical issue.

**Comment by batmat** [comment]
Great you're trying to address an issue and align/unify our plugins management policies! Thanks for the effort!

If you deem this subject is critical, please please raise a discussion on the dev ML... As a rule, anything really debatable needs, well, debate with the community. And that happens there, normally not in a PR (possibly a JEP candidate BTW).

Now you may think it's not debatable, but the comments here definitely seem to imply we need more open ended discussion before reaching any solution.
(And gov meeting ack should happen at the end of this. Not before. When things are taken to the gov meeting w/o prior debate, the subject is almost always pushed back to a thread on the ML...)

**Comment by recena** [comment]
@batmat you are totally right! Thanks

**Comment by batmat** [comment]
Closing as a followup of my last comment. I think a JEP is needed here if we still want to move forward. Also the upstream PR in extra-enforcer-rules is not merged either.

Please feel free to reopen/comment when/if you'd like to get back to this.

Thank you!

**Pull request closed without merging**